src/session/content/background.rs

```diff
@@ -0,0 +1,409 @@
+        pub(super) pattern: OnceCell<gdk::Texture>,
+
+        pub(super) progress: Cell<f32>,
+        pub(super) phase: Cell<u32>,
+
+        pub(super) dark: Cell<bool>,
+    }
+
+    #[glib::object_subclass]
+    impl ObjectSubclass for Background {
+        const NAME: &'static str = "ComponentsBackground";
+        type Type = super::Background;
+        type ParentType = gtk::Widget;
+    }
+
+    impl ObjectImpl for Background {
+        fn constructed(&self) {
+            self.parent_constructed();
+
+            let obj = self.obj();
+
+            obj.set_vexpand(true);
+            obj.set_hexpand(true);
+
+            let pattern =
+                gdk::Texture::from_resource("/com/github/melix99/telegrand/images/pattern.svg");
+
+            self.pattern.set(pattern).unwrap();
+
+            let style_manager = adw::StyleManager::default();
+            self.dark.set(style_manager.is_dark());
+
+            style_manager.connect_dark_notify(clone!(@weak obj => move |style_manager| {
+                let imp = obj.imp();
+                imp.dark.set(style_manager.is_dark());
+                imp.gradient_cache.take();
+            }));
+
+            let target = adw::CallbackAnimationTarget::new(clone!(@weak obj => move |progress| {
+                let imp = obj.imp();
+                imp.gradient_cache.take();
+                let progress = progress as f32;
+                if progress >= 1.0 {
+                    imp.progress.set(0.0);
+                    imp.phase.set((imp.phase.get() + 1) % 8);
+                } else {
+                    imp.progress.set(progress)
+                }
+                obj.queue_draw();
+            }));
+
+            let animation = adw::TimedAnimation::new(obj.as_ref(), 0.0, 1.0, 200, &target);
+            animation.set_easing(adw::Easing::EaseInOutQuad);
+
+            obj.connect_local("animate", true, move |_| {
+                let val = animation.value();
+                if val == 0.0 || val == 1.0 {
+                    animation.play()
+                }
+                None
+            });
+        }
+
+        fn signals() -> &'static [Signal] {
+            static SIGNALS: Lazy<Vec<Signal>> =
+                Lazy::new(|| vec![Signal::builder("animate").return_type::<()>().build()]);
+
+            SIGNALS.as_ref()
+        }
+    }
+
+    impl WidgetImpl for Background {
+        fn measure(&self, orientation: gtk::Orientation, for_size: i32) -> (i32, i32, i32, i32) {
+            if let Some(child) = self.obj().first_child() {
+                child.measure(orientation, for_size)
+            } else {
+                self.parent_measure(orientation, for_size)
+            }
+        }
+
+        fn size_allocate(&self, width: i32, height: i32, baseline: i32) {
+            if let Some(child) = self.obj().first_child() {
+                child.allocate(width, height, baseline, None)
+            }
+        }
```

Comment:
This widget really feels like an `AdwBin` at this point. I suggest subclassing from that and remove these functions.